cpu/lm4f120/periph/spi.c

```diff
@@ -0,0 +1,271 @@
+    ROM_GPIOPinConfigure(spi_confs[dev].pins.tx);
+
+    ROM_GPIOPinTypeSSI(spi_confs[dev].gpio_port, spi_confs[dev].pins.mask);
+
+    switch (speed) {
+        case SPI_SPEED_100KHZ:
+            bitrate = 100000;
+            break;
+        case SPI_SPEED_1MHZ:
+            bitrate = 1000000;
+            break;
+        case SPI_SPEED_5MHZ:
+            bitrate = 5000000;
+            break;
+        case SPI_SPEED_10MHZ:
+            bitrate = 10000000;
+            break;
+        case SPI_SPEED_400KHZ:
+        default:
+            bitrate = 400000;
+            break;
+    }
+
+    /*
+       Polarity Phase        Mode
+       0     0   SSI_FRF_MOTO_MODE_0
+       0     1   SSI_FRF_MOTO_MODE_1
+       1     0   SSI_FRF_MOTO_MODE_2
+       1     1   SSI_FRF_MOTO_MODE_3
+     */
+    switch (conf) {
+        case SPI_CONF_FIRST_RISING: /**< first data bit is transacted on the first rising SCK edge */
+            mode = SSI_FRF_MOTO_MODE_0;
+            break;
+        case SPI_CONF_SECOND_RISING: /**< first data bit is transacted on the second rising SCK edge */
+            mode = SSI_FRF_MOTO_MODE_1;
+            break;
+        case SPI_CONF_FIRST_FALLING: /**< first data bit is transacted on the first falling SCK edge */
+            mode = SSI_FRF_MOTO_MODE_2;
+            break;
+        case SPI_CONF_SECOND_FALLING: /**< first data bit is transacted on the second falling SCK edge */
+            mode = SSI_FRF_MOTO_MODE_3;
+            break;
+    }
+
+    ROM_SSIClockSourceSet(spi_confs[dev].ssi_base, SSI_CLOCK_SYSTEM);
+    ROM_SSIConfigSetExpClk(spi_confs[dev].ssi_base, ROM_SysCtlClockGet(),
+                           mode,
+                           SSI_MODE_MASTER,
+                           bitrate,
+                           8);
+
+    ROM_SSIEnable(spi_confs[dev].ssi_base);
+
+    while (ROM_SSIDataGetNonBlocking(spi_confs[dev].ssi_base, &initialData)) ;
+    return 0;
+}
+
+int spi_init_slave(spi_t dev, spi_conf_t conf, char (*cb)(char))
+{
+    return -1;
+}
+
+void spi_transmission_begin(spi_t dev, char reset_val)
+{
+}
+
+int spi_acquire(spi_t dev)
+{
+    if (dev >= SPI_NUMOF) {
+        return -1;
+    }
+    mutex_lock(&locks[dev]);
+    return 0;
+}
+
+int spi_release(spi_t dev)
+{
+    if (dev >= SPI_NUMOF) {
+        return -1;
+    }
+    mutex_unlock(&locks[dev]);
+    return 0;
+}
+
+int spi_transfer_byte(spi_t dev, char out, char *in)
+{
+    unsigned long lin;
+
+    if (dev >= SPI_NUMOF) {
+        return -1;
+    }
+    DEBUG("SPI byte transfer on %d (before in: %x, out: %x) ", dev, *in, out);
+    ROM_SSIDataPut(spi_confs[dev].ssi_base, out);
+
+    // wait until tx over
+    while (ROM_SSIBusy(spi_confs[dev].ssi_base)) ;
+
+    ROM_SSIDataGet(spi_confs[dev].ssi_base, &lin);
+
+    // wait until rx over
+    while (ROM_SSIBusy(spi_confs[dev].ssi_base)) ;
+
+    *in = (char)lin;
+    DEBUG("in %x, out %x\n", *in, out);
+
+    return 0;
+}
+
+void spi_poweron(spi_t dev)
+{
+    if (dev >= SPI_NUMOF) {
```

Comment:
If I got it right, the setup above is allowing arbitrary device settings ( e.g. `SPI_NUMOF=2` with `SPI_0_EN, SPI_3_EN` defined). This check wouldn't handle that correctly.